colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,290 @@
+                if p.grad is None:
+                    continue
+                
+                # ==============================
+                # Init Current Param Shape
+                # ==============================
+                p_shape = None  # param shape (not split)
+                param_height = None
+                param_width = None
+                param_height_parallel = None
+                param_width_parallel = None
+                sharding_spec = None
+                if id(p) in self.param_shape.keys():
+                    p_shape = self.param_shape[id(p)]
+                    # if id(p) in self.sharding_spec_dict.keys() and self.sharding_spec_dict[id(p)] != None:
+                    #     sharding_spec = self.sharding_spec_dict[id(p)]
+                    if len(p_shape) >= 2: # factored
+                        param_height = p_shape[0]
+                        param_width = p_shape[1]
+                        sharding_spec = self.sharding_spec_dict[id(p)]
+                        if sharding_spec.sharding_sequence[0] == 'R': # Col Parallel 
+                            param_height_parallel = param_height
+                            param_width_parallel = param_width // self.tensor_parallel_size # W/N
+                        if sharding_spec.sharding_sequence[-1] == 'R': # Row Parallel
+                            param_height_parallel = param_height // self.tensor_parallel_size # H/N
+                            param_width_parallel = param_width
+                
+                # grad shape is same as weigh / bias
+                grad = p.grad.to(self.localRank)
+                if grad.dtype in {torch.float16, torch.bfloat16}:
+                    grad = grad.float()
+                if grad.is_sparse:
+                    raise RuntimeError("Adafactor does not support sparse gradients.")
+                
+                """
+                p (group["params"]) is weight
+                state 
+                {'step', 
+                'exp_avg_sq_row', 
+                'exp_avg_sq_col', 
+                'RMS'
+                }
+                
+                p (group["params"]) is bias
+                state 
+                {'step', 
+                'exp_avg_sq', 
+                'RMS'
+                }
+                """
+                state = self.state[p]  # always empty
+                # grad_shape = grad.shape
+                factored, use_first_moment = self._get_options(group, p_shape)
+                if len(state) == 0:
+                    state["step"] = 0
+                    if use_first_moment:
+                        # Exponential moving average of gradient values
+                        state["exp_avg"] = torch.zeros_like(grad)
+                    if factored:
+                        if sharding_spec.sharding_sequence[0] == 'R': # Col Parallel
+                            state["exp_avg_sq_row"] = torch.zeros(param_height).to(grad)  # [H:4096]
```

Comment:
`torch.zeros(param_height, device=grad.device)`

Reply:
> torch.zeros(param_height, device=grad.device)
Modify set tensor device "torch.zeros(param_height).to(grad)" --> torch.zeros(param_height, device=grad.device);
Modified in: https://github.com/hpcaitech/ColossalAI/pull/5484/commits/db56b2f903f3ab430ec8cefe479d39d6eba466e6